### ISSUE TITLE
[BUG] Ableton Live crashes when opening projects containing Chowtape VST3

### ISSUE THREAD

**Issue by mstgrv** [jatinchowdhury18/AnalogTapeModel#64]
Ableton Live is crashing when trying to open existing projects that contain VST3 instances of Chowtape. I'm getting consistent crashes across multiple projects - one is a super simple / small project, the other is quite large.

Creating a fresh instance of Chowtape VST3 in an empty project does not crash though, it only seems to happen when opening an existing project which contains Chowtape VST3.

Chowtape 2.4.0 VST3
macOS 10.14.6
Ableton Live 10.1.17 and 10.1.15

Crash report generated by macOS:

Thread 0 Crashed:: MainThread  Dispatch queue: com.apple.main-thread
0   com.ableton.live              	0x0000000100fc79a8 0x100000000 + 16546216
1   com.ableton.live              	0x0000000100fd8a67 0x100000000 + 16616039
2   com.ableton.live              	0x000000010257d754 0x100000000 + 39311188
3   com.ableton.live              	0x0000000102573948 0x100000000 + 39270728
4   com.ableton.live              	0x0000000102573a3d 0x100000000 + 39270973
5   com.chowdsp.CHOWTapeModel     	0x000000018ed37af7 ChowtapeModelAudioProcessor::prepareToPlay(double, int) + 6343
6   com.chowdsp.CHOWTapeModel     	0x000000018ed10e91 juce::JuceVST3Component::setActive(unsigned char) + 769
7   com.ableton.live              	0x000000010239f835 0x100000000 + 37353525
8   com.ableton.live              	0x000000010239ca2e 0x100000000 + 37341742
9   com.ableton.live              	0x000000010239b19e 0x100000000 + 37335454
10  com.ableton.live              	0x00000001023960b9 0x100000000 + 37314745
11  com.ableton.live              	0x0000000101387aee 0x100000000 + 20478702
12  com.ableton.live              	0x0000000101387898 0x100000000 + 20478104
13  com.ableton.live              	0x0000000101398d9e 0x100000000 + 20549022
14  com.ableton.live              	0x000000010136ad98 0x100000000 + 20360600
15  com.ableton.live              	0x0000000101399556 0x100000000 + 20550998
16  com.ableton.live              	0x000000010194f8a5 0x100000000 + 26540197
17  com.ableton.live              	0x000000010194f72a 0x100000000 + 26539818
18  com.ableton.live              	0x0000000101939868 0x100000000 + 26450024
19  com.ableton.live              	0x0000000101939bed 0x100000000 + 26450925
20  com.ableton.live              	0x0000000101932244 0x100000000 + 26419780
21  com.ableton.live              	0x0000000101959bc0 0x100000000 + 26581952
22  com.ableton.live              	0x000000010195972d 0x100000000 + 26580781
23  com.ableton.live              	0x00000001019588a8 0x100000000 + 26577064
24  com.ableton.live              	0x0000000101c572d1 0x100000000 + 29717201
25  com.ableton.live              	0x0000000101e399f2 0x100000000 + 31693298
26  com.ableton.live              	0x0000000101e50cb6 0x100000000 + 31788214
27  com.ableton.live              	0x0000000101edb3fd 0x100000000 + 32355325
28  com.ableton.live              	0x000000010057b3c9 0x100000000 + 5747657
29  com.ableton.live              	0x00000001005a3736 0x100000000 + 5912374
30  com.ableton.live              	0x00000001011e35ab 0x100000000 + 18757035
31  com.ableton.live              	0x00000001011e46b8 0x100000000 + 18761400
32  com.ableton.live              	0x00000001010e2e9b 0x100000000 + 17706651
33  com.ableton.live              	0x00000001011e3b5d 0x100000000 + 18758493
34  com.ableton.live              	0x000000010117f6e1 0x100000000 + 18347745
35  com.ableton.live              	0x00000001010c1a4c 0x100000000 + 17570380
36  com.apple.AppKit              	0x00007fff510a1644 -[NSApplication(NSResponder) sendAction:to:from:] + 312
37  com.apple.AppKit              	0x00007fff510fd86b -[NSMenuItem _corePerformAction] + 323
38  com.apple.AppKit              	0x00007fff510fd5dc -[NSCarbonMenuImpl performActionWithHighlightingForItemAtIndex:] + 114
39  com.apple.AppKit              	0x00007fff5116158f -[NSMenu performActionForItemAtIndex:] + 131
40  com.apple.AppKit              	0x00007fff511614fc -[NSMenu _internalPerformActionForItemAtIndex:] + 93
41  com.apple.AppKit              	0x00007fff5116132e -[NSCarbonMenuImpl _carbonCommandProcessEvent:handlerCallRef:] + 105
42  com.apple.AppKit              	0x00007fff510cbdc6 NSSLMMenuEventHandler + 909
43  com.apple.HIToolbox           	0x00007fff52a1622e DispatchEventToHandlers(EventTargetRec*, OpaqueEventRef*, HandlerCallRec*) + 1422
44  com.apple.HIToolbox           	0x00007fff52a155df SendEventToEventTargetInternal(OpaqueEventRef*, OpaqueEventTargetRef*, HandlerCallRec*) + 371
45  com.apple.HIToolbox           	0x00007fff52a329b6 SendEventToEventTarget + 39
46  com.apple.HIToolbox           	0x00007fff52a84cd2 SendHICommandEvent(unsigned int, HICommand const*, unsigned int, unsigned int, unsigned char, void const*, OpaqueEventTargetRef*, OpaqueEventTargetRef*, OpaqueEventRef**) + 375
47  com.apple.HIToolbox           	0x00007fff52aad14e SendMenuCommandWithContextAndModifiers + 49
48  com.apple.HIToolbox           	0x00007fff52aad0fd SendMenuItemSelectedEvent + 192
49  com.apple.HIToolbox           	0x00007fff52aacfe5 FinishMenuSelection(SelectionData*, MenuResult*, MenuResult*) + 96
50  com.apple.HIToolbox           	0x00007fff52aad9b3 MenuSelectCore(MenuData*, Point, double, unsigned int, OpaqueMenuRef**, unsigned short*) + 603
51  com.apple.HIToolbox           	0x00007fff52aad6c8 _HandleMenuSelection2 + 452
52  com.apple.AppKit              	0x00007fff50db9a81 _NSHandleCarbonMenuEvent + 223
53  com.apple.AppKit              	0x00007fff50db98e6 _DPSEventHandledByCarbon + 54
54  com.apple.AppKit              	0x00007fff50dae92b -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 2577
55  com.apple.AppKit              	0x00007fff50da8588 -[NSApplication run] + 699
56  com.ableton.live              	0x00000001010be76c 0x100000000 + 17557356
57  libdyld.dylib                 	0x00007fff7f75d3d5 start + 1

**Comment by mstgrv** [comment]
Seems the issue is related to oversampling:

1. Create a new Live set (in Live 10.1.17)
2. Create an instance of Chowtape VST3 and set oversampling to 16x
3. Save the project
4. Reopen the project - Live will crash
5. Opening the project in Live 10.1.1 - it will open without any problems.

AU and VST2 versions of the plugin don't have this issue.

**Comment by jatinchowdhury18** [comment]
Hi @mstgrv,

Thanks for the very detailed bug report! Very strange that it's only happening on the VST3 version...

I should have some time to check it out over the next couple days, hopefully the issue re-creates correctly on Windows, as I don't currently have a Mac with Ableton available for testing. In the meantime, here's a couple other experiments you could try that might be useful for debugging this:

1. Try with different sample rates / buffer sizes set in Ableton. Any different behavior at different rates and buffer sizes would be useful to know.

2. Try with an older version of CHOW Tape. I'd be curious to know if version 2.3.4 has the same behaviour.

3. Try with different hysteresis mode settings. I'm pretty doubtful that this would make a difference, but given that the parameters are somewhat connected, there is a possibility that the issue is linked to both parameters.

Sorry for the inconvenience, hopefully we can get this fixed pretty soon.

Thanks,
Jatin

**Comment by mstgrv** [comment]
Haven't been able to narrow this down much at all, but it's definitely happening at both 44.1k and 48k sample rates. Will post back if I find any further details

**Comment by jatinchowdhury18** [comment]
Hi @mstgrv,

Thanks for continuing to look into this. If you have a minute, could you try again with some new builds here: https://github.com/jatinchowdhury18/AnalogTapeModel/blob/vst3-load-bug/Plugin/Bin/MacBuilds.zip. I've tweaked a couple things that I think should help.

Also, if you would be able to share one of your sessions that has been having this bug (preferably the smaller one), that would be helpful.

Thanks,
Jatin

**Comment by mstgrv** [comment]
Thanks Jatin - this seems to work fine so far without crashing. Here's my extremely basic test setup:

1. Start a new Ableton project
2. Create a VST3 instance of Chowtape VST3 and set oversampling to 16x (don't touch anything else)
3. Save the project
4. Reopen the project

With Chowtape 2.4.0, Live crashes at step 4. With the new build you just posted, it will reopen successfully without crashing.

**Comment by jatinchowdhury18** [comment]
Hi @mstgrv,

Thanks for the update. Unfortunately, I've been unable to recreate this crash on my Windows machine. Hoping I can get a chance to debug directly on a Mac soon...

In the meantime, the fact that the updated version seems to work without crashing is a good sign. I had disabled a couple things in that version that I figured might be the cause of the issues, so I've made a new version that re-enables those things in a different way. If you have a minute to test this new version that would be super helpful: https://github.com/jatinchowdhury18/AnalogTapeModel/blob/vst3-load-bug/Plugin/Bin/MacBuilds.zip.

Thanks,
Jatin
